### DISCUSSION TITLE
"No refactorings available" / "No code actions" when cursor is inside of a string

### DISCUSSION THREAD

**Discussion by DamianFekete** [microsoft/pylance-release#6344] [Enhancement]
Type: <b>Bug</b>

# Behaviour

No refactorings available / No code actions for 'refactor.extract.variable' available

## Steps to reproduce:

* Create a new folder
* Create a file called testvscode.py with the following content

```python
def f():
    print("Ohh no")
    
f()
```

* Open the folder in VS Code
* Right click somewhere in the text `Ohh no`. Choose `Refactor...` from the popup menu. The message is: _No refactorings available._
    I think it should allow you to at least extract the text as a variable.
![358513518-2250296f-8293-4831-a024-9f17271a63bc](https://github.com/user-attachments/assets/0683f48e-4733-4459-ad8e-75d65c6896fb)

* If I use the keyboard shortcut to extract text to variable I get this error:
![358513704-cb9d250d-f394-46cd-8657-7edf5a4e071c](https://github.com/user-attachments/assets/97b93071-87ce-4541-aa22-bfa10a5068fb)

* I have tried using not only the release version of the Python/Pylance plugins, but the pre-release versions as well.
The default Language Server for Python was default, I have set it explicitly to Pylance now, without success:
![358514034-bc0f9d26-e50f-4ff1-884f-069e96e33d45](https://github.com/user-attachments/assets/db3f9db9-6109-4e7d-ac73-4c47e1eaf7f1)

* I don't see any other errors and don't know how to debug this issue further.
![358514595-a4b9495b-2c7f-4a44-99e4-d4440baee060](https://github.com/user-attachments/assets/23dc3e0e-c9c0-49d6-912c-5406afe706d8)

# Diagnostic data

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
XXX
```

</p>
</details>

Extension version: 2024.12.3
VS Code version: Code 1.92.2 (fee1edb8d6d72a0ddff41e5f71a671c23ed924b9, 2024-08-14T17:29:30.058Z)
OS version: Windows_NT x64 10.0.19045
Modes:

-   Python version (& distribution if applicable, e.g. Anaconda): 3.11.2
-   Type of virtual environment used (e.g. conda, venv, virtualenv, etc.): Global
-   Value of the `python.languageServer` setting: Pylance

<details>
<summary>User Settings</summary>
<p>

```

languageServer: "Pylance"

```
</p>
</details>

<details>
<summary>Installed Extensions</summary>

|Extension Name|Extension Id|Version|
|---|---|---|
|Edit csv|janisdd.vscode-edit-csv|0.9.2|
|EditorConfig for VS Code|EditorConfig.EditorConfig|0.16.4|
|IntelliJ IDEA Keybindings|k--kato.intellij-idea-keybindings|1.7.2|
|JavaScript Debugger|ms-vscode.js-debug|1.92.0|
|JavaScript Debugger Companion Extension|ms-vscode.js-debug-companion|1.1.3|
|PowerShell|ms-vscode.powershell|2024.2.2|
|Pylance|ms-python.vscode-pylance|2024.8.2|
|Python|ms-python.python|2024.12.3|
|Python Debugger|ms-python.debugpy|2024.11.2024082901|
|SailPoint Identity Security Cloud|yannick-beot-sp.vscode-sailpoint-identitynow|1.3.2|
|Table Visualizer for JavaScript Profiles|ms-vscode.vscode-js-profile-table|1.0.9|
</details>
<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz (16 x 2304)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|31.73GB (12.73GB free)|
|Process Argv|.|
|Screen Reader|no|
|VM|0%|
</details>

The original bug report against VS Code was closed: https://github.com/microsoft/vscode/issues/225766

**Comment by rchiodo** [comment]
Thanks for the issue. You have to select the text for this to work (otherwise we don't know what to select to extract).



https://github.com/user-attachments/assets/bbb1f1dd-3d35-4a50-84e6-03784ed38f0a

We could potentially special case strings to auto pick the whole thing, but there's a perf penalty with generating code actions everywhere.

**Reply by DamianFekete** [reply]
As I am using mostly IntelliJ I did not expect that I have to manually select the text.
The same thing for naming numbers.
For comparison:
![image](https://github.com/user-attachments/assets/d469cb4f-7865-4350-a9e2-26f27ae6e23a)


> there's a perf penalty with generating code actions everywhere

Is the performance penalty when you load/process the file or when I press the key? If it is only when I press the key, the computer should be faster than me to increase selection and find a suitable element (string, number), that can be refactored.

**Comment by rchiodo** [comment]
Reopening. Going to mark as an enhancement request for upvotes.
